torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -764,44 +784,131 @@ def _device_to_params_per_rank(
+    def _get_min_index(self, values: List[int]) -> int:
         r"""
-        :class:`dict` mapping each device to a :class:`list` of :class:`list`
-        of :class:`_DDPBucket` s.
+        Returns ``values.index(min(values))``, except only uses one pass.
 
-        ``_device_to_buckets[d][r][i]`` gives the ``i``th bucket
-        assigned to rank ``r`` stored on device ``d``, where each bucket
-        contains a list of the model parameters associated with the
-        corresponding logical :class:`DistributedDataParallel` gradient bucket.
+        Arguments:
+            values: (List[int]): :class:`list` of values.
+        """
+        min_index = 0
+        min_value = float("inf")
+        for i, value in enumerate(values):
+            if value < min_value:
+                min_value = value
+                min_index = i
+        return min_index
+
+    def _assign_bucket_subset_to_rank(
+        self,
+        bucket_index: int,
+        bucket_params: List[torch.Tensor],
+        assigned_ranks_per_bucket: List[Set[int]],
+        assignment_size: int,
+        size_per_rank: List[int],
+    ) -> None:
+        r"""
+        Assigns the model parameters given by ``bucket_params``, representing a
+        (possibly non-strict) subset of the parameters corresponding to a
+        :class:`DistributedDataParallel` bucket, to the rank with the least
+        size assigned so far and collects relevant information.
 
-        This is used for constructing the parameter buckets if
-        ``overlap_with_ddp=True``.
+        Arguments:
+            bucket_index (int): index of the :class:`DistributedDataParallel`
+                gradient bucket.
+            bucket_params (List[torch.Tensor]): subset of the parameters
+                corresponding to the bucket to assign.
+            assigned_ranks_per_bucket (List[Set[int]]): :class:`set` of ranks
+                assigned to each bucket.
+            assignment_size (int): assignment size (i.e. sum of
+            ``param.numel()`` for all ``param`` in ``bucket_params``).
+            size_per_rank (List[int]): size so far for each rank.
         """
-        assert self._overlap_with_ddp, \
-            "`_device_to_buckets()` should only be used if " \
-            "`overlap_with_ddp=True`"
-        if len(self._device_to_buckets_cache) > 0:
-            return self._device_to_buckets_cache
+        overlap_info = self._overlap_info
+        assert overlap_info.needs_global_bucket_info, \
+            "`_assign_bucket_subset_to_rank` should only be used if " \
+            "`needs_global_bucket_info=True`"
+        params_per_rank = overlap_info.params_per_rank
+        offsets = overlap_info.offsets
+        assigned_rank = self._get_min_index(size_per_rank)
+
+        self._bucket_assignments_per_rank_cache[assigned_rank][bucket_index] = \
+            _DDPBucketAssignment(bucket_index, bucket_params)
+        # NOTE: The following matches the information collected in
```

Comment:
It may be possible to refactor this and its duplicate in `_collect_ddp_bucket_info()`. If you think it is worth refactoring, maybe I make it into a method in `_OverlapInfo`?

Reply:
I think one way to make this simpler may be to also delay the information collection to after all buckets have been seen for `hook_with_zero_step(uniform_bucket_assignment=False)` (i.e. Approach 2) instead of doing so as each bucket is being seen. There should be no difference semantically; the ZeRO instance will still be fully initialized at the end of the second iteration.

Reply:
Refactored this, so marking as resolved.